clippy_lints/src/tabs_in_doc_comments.rs

```diff
@@ -0,0 +1,198 @@
+use crate::utils::span_lint_and_sugg;
+use rustc::lint::{EarlyContext, EarlyLintPass, LintArray, LintPass};
+use rustc::{declare_lint_pass, declare_tool_lint};
+use rustc_errors::Applicability;
+use std::convert::TryInto;
+use syntax::ast::{self};
+use syntax::source_map::BytePos;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks doc comments for usage of tab characters.
+    ///
+    /// **Why is this bad?** The rust style-guide promotes spaces instead of tabs for indentation.
+    /// To keep a consistent view on the source, also doc comments should not have tabs.
+    /// Also, explaining ascii-diagrams containing tabs can get displayed incorrectly when the
+    /// display settings of the author and reader differ.
+    ///
+    /// **Known problems:** None on the lint. If using rustfix, all tabs will be replaced by four
+    /// spaces. This may corrupt tab-aligned text.
+    ///
+    /// **Example:**
+    /// ```rust,ignore
+    /// ///
+    /// /// Struct to hold two strings:
+    /// /// 	- first		one
+    /// /// 	- second	one
+    /// pub struct DoubleString {
+    ///    ///
+    ///    /// 	- First String:
+    ///    /// 		- needs to be inside here
+    ///    first_string: String,
+    ///    ///
+    ///    /// 	- Second String:
+    ///    /// 		- needs to be inside here
+    ///    second_string: String,
+    ///}
+    /// ```
+    ///
+    /// Will be converted to:
+     /// ```rust,ignore
+    /// ///
+    /// /// Struct to hold two strings:
+    /// ///     - first        one
+    /// ///     - second    one
+    /// pub struct DoubleString {
+    ///    ///
+    ///    ///     - First String:
+    ///    ///         - needs to be inside here
+    ///    first_string: String,
+    ///    ///
+    ///    ///     - Second String:
+    ///    ///         - needs to be inside here
+    ///    second_string: String,
+    ///}
+    /// ```
+    pub TABS_IN_DOC_COMMENTS,
+    style,
+    "Using tabs in doc comments is not recommended. Consider using four spaces per tab."
+}
+
+declare_lint_pass!(TabsInDocComments => [TABS_IN_DOC_COMMENTS]);
+
+impl TabsInDocComments {
+    fn warn_if_tabs_in_doc(cx: &EarlyContext<'_>, attr: &ast::Attribute) {
+        if let ast::AttrKind::DocComment(comment) = attr.kind {
+            let comment = comment.as_str();
+
+            for (lo, hi) in get_chunks_of_tabs(&comment) {
+                let new_span = attr
+                    .span
+                    .with_lo(attr.span.lo() + BytePos(lo.try_into().unwrap_or(u32::max_value())))
+                    .with_hi(attr.span.lo() + BytePos(hi.try_into().unwrap_or(u32::max_value())));
```

Comment:
This is my main concern:

If for some reason `usize`->`u32` fails, `new_span` will have weird and unpredictable bounds (and Clippy would ICE with a weird message). I would just return `u32`s, to hide possibly necessary conversions. 

If you decide to keep returning `usize`s, I'd prefer to use `expect("...")` here, so when Clippy should ICE from this, it's easier to find.

---
Also, you're creating a completely new span here, so I'd suggest to just use the `Span::new` function for this:
```rust
let new_span = Span::new(
    attr.span.lo() + BytePos(lo),
    attr.span.lo() + BytePos(hi),
    attr.span.ctxt(),
);
```

Reply:
Fair point. Taking it up!
Ah, thanks for the 'attr.span.ctxt()', we were missing that :D